ROX-28148: Collect metrics for scan semaphore queue sizes

jschnath opened the pull request:
### Description

Adds a prometheus metric that tracks the scan semaphore queue size in image scan.

### User-facing documentation

- [ ] CHANGELOG is updated **OR** update is not needed
- [ ] [documentation PR](https://spaces.redhat.com/display/StackRox/Submitting+a+User+Documentation+Pull+Request) is created and is linked above **OR** is not needed

### Testing and quality

<!--
General Availability requirements: https://github.com/stackrox/stackrox/blob/master/PR_GA.md
Feature Flags usage: https://github.com/stackrox/stackrox/blob/master/pkg/features/README.md
-->

- [ ] the change is production ready: the change is GA or otherwise the functionality is gated by a feature flag
- [ ] CI results are inspected

#### Automated testing

<!--
If no tests have been contributed, please explain why unless it's obvious,
e.g., the PR is a one-line comment change.
-->

- [ ] added unit tests
- [ ] added e2e tests
- [ ] added regression tests
- [ ] added compatibility tests
- [ ] modified existing tests

#### How I validated my change

Not sure if it makes sense to add automated tests here(?) For other simple metrics we don't have tests at least. If automated tests are not required I suppose the best way to test this is to put the image on a cluster and trigger image scans to see if the metric reports expected results, but I have not yet done that.

dcaravel commented on central/image/service/service_impl.go:
```diff
@@ -476,9 +477,15 @@ func (s *serviceImpl) GetImageVulnerabilitiesInternal(ctx context.Context, reque
 func (s *serviceImpl) acquireScanSemaphore(ctx context.Context) error {
 	semaphoreCtx, cancel := context.WithTimeout(ctx, maxSemaphoreWaitTime)
 	defer cancel()
+	metrics.ImageScanSemaphoreQueueSize.Add(1)
 	if err := s.internalScanSemaphore.Acquire(semaphoreCtx, 1); err != nil {
 		wrappedErr := errors.Wrap(err, "acquiring scan semaphore")
```
Consider incrementing the count after `Acquire` is successful.

In the scenario where the Acquire was unsuccessful, the count will be increased but not decreased (`s.releaseFromScanSemaphore(1)` would not be called)

jschnath replied:
I've refactored the whole metric usage, I think the first one I tried made little sense.

dcaravel replied:
Ahh - I see now,  the intent was to capture # of requests waiting.

stehessel reviewed (commented):
I think we probably want to define more than one metric to get the full picture. Some inspiration:

* How many requests are currently waiting on the semaphore?
* How many requests successfully hold the semaphore?
* What is the total semaphore limit?

Also note that there are a few different scan related semaphores (see local scanner code in sensor and the enricher in Central).

jschnath commented:
@stehessel Thanks for the comment, I think I understand how we use these metrics a little bit better now. I've split them up to give us the data you want and added metrics to the enricher and sensor/scan as well.

Note that the semaphore limit is hard to track because it's a private field of the semaphore set during ctor and cannot be queried.

jschnath commented:
/test help

jschnath commented:
/test gke-perf-scale-tests

vikin91 commented on central/image/metrics/metrics.go:
```diff
@@ -0,0 +1,33 @@
+		Namespace: metrics.PrometheusNamespace,
+		Subsystem: metrics.CentralSubsystem.String(),
+		Name:      "image_scan_semaphore_queue_size",
+		Help:      "A counter that tracks the size of the queue for the scan semaphore used in image scan.",
```
```suggestion
		Help:      "A gauge that shows the size of the queue for the scan semaphore used in image scan.",
```

vikin91 commented on central/image/metrics/metrics.go:
```diff
@@ -0,0 +1,33 @@
+		Namespace: metrics.PrometheusNamespace,
+		Subsystem: metrics.CentralSubsystem.String(),
+		Name:      "image_scan_semaphore_holding_size",
+		Help:      "A counter that tracks the number of requests successfully holding the semaphore.",
```
```suggestion
		Help:      "A gauge that tracks the number of requests successfully holding the semaphore.",
```

vikin91 commented on central/image/metrics/metrics.go:
```diff
@@ -0,0 +1,33 @@
+		Namespace: metrics.PrometheusNamespace,
+		Subsystem: metrics.CentralSubsystem.String(),
+		Name:      "image_scan_semaphore_limit",
+		Help:      "A counter that tracks the limit of available scan semaphores.",
```
```suggestion
		Help:      "A gauge that tracks the limit of available scan semaphores.",
```

vikin91 commented on central/image/metrics/metrics.go:
```diff
@@ -0,0 +1,33 @@
+)
+
+func init() {
+	prometheus.MustRegister(ImageScanSemaphoreQueueSize,
+		ImageScanSemaphoreHoldingSize,
+		ImageScanSemaphoreLimit)
```
How about we combine those 3 gauges into one GaugeVec and use 3 labels: queueSize, holdingSize, limit?

jschnath replied:
So ideally I would want three metrics then,

- ImageScanSemaphoreQueueSize
- ImageScanSemaphoreHoldingSize
- ImageScanSemaphoreLimit

But make them of the type `GaugeVec` with a label "location" denoting where something is from, e.g. "image scan"? But can this work when the metrics are pulled from completely separate locations (central, sensor, some package)?

vikin91 replied:
This looks addressed now.

vikin91 commented on pkg/images/enricher/metrics.go:
```diff
@@ -72,13 +95,27 @@ func newMetrics(subsystem pkgMetrics.Subsystem) metrics {
+			Namespace: pkgMetrics.PrometheusNamespace,
+			Subsystem: subsystem.String(),
+			Name:      "enricher_semaphore_queue_size",
+			Help:      "A counter that tracks the size of the queues for the scan semaphores used in image scans.",
```
This is not a counter!

vikin91 replied:
Those two extra metrics could also be another label for the previous metric. Something like: `location=enricher`, whereas in the other case it would be `location=image-scan`. That would make it more compact and easier to find in the list of metrics.

That comment with labels make sense only for the metrics that are in the same origin: i.e., one metric with labels for Sensor, another metric with labels for Central.

jschnath replied:
> Those two extra metrics could also be another label for the previous metric. Something like: location=enricher, whereas in the other case it would be location=image-scan

I like this idea but I'm having some trouble implementing this. The 3 sets of metrics I'm adding are in very different parts of the code, one in central, one in sensor, one in a package. Is it really possible to combine them all into a few `GaugeVec`s or am I misunderstanding what you're saying I should try?

stehessel replied:
You can define the metrics in different places as long as they have the same name. Just make sure to separate them by label.

vikin91 replied:
I think, this has been addressed by Jan. Resolving.

vikin91 commented on pkg/images/enricher/metrics.go:
```diff
@@ -72,13 +95,27 @@ func newMetrics(subsystem pkgMetrics.Subsystem) metrics {
+			Namespace: pkgMetrics.PrometheusNamespace,
+			Subsystem: subsystem.String(),
+			Name:      "enricher_semaphore_holding_size",
+			Help:      "A counter that tracks the number of requests successfully holding scanner semaphores",
```
Also not a counter

vikin91 commented on sensor/common/scan/metrics/metrics.go:
```diff
@@ -0,0 +1,26 @@
+	SensorScanSemaphoreQueueSize = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: metrics.PrometheusNamespace,
+		Subsystem: metrics.SensorSubsystem.String(),
+		Name:      "sensor_scan_semaphore_queue_size",
```
Same comment about the label here: `location=sensor-scan`, although I ask myself what is "sensor-scan"? Maybe a better name here would be useful.

vikin91 replied:
I went with longer `sensor-local-image-scan`

vikin91 commented on sensor/common/scan/metrics/metrics.go:
```diff
@@ -0,0 +1,26 @@
+		Namespace: metrics.PrometheusNamespace,
+		Subsystem: metrics.SensorSubsystem.String(),
+		Name:      "sensor_scan_semaphore_queue_size",
+		Help:      "A counter that tracks the size of the queue for the scan semaphore used in sensor scan.",
```
This is not a counter! Please check for this text globally and improve

vikin91 reviewed (changes requested):
I left few comments about the design of metrics and the help text. Having too many separate metrics is something that we should rather avoid, unless there is a reason for that.

stehessel commented on central/image/service/service_impl.go:
```diff
@@ -86,6 +88,8 @@ var (
 
 	delegateScanPermissions = []string{"Image"}
+
+	imageScanMetricsName = prometheus.Labels{"location": "image-scan"}
```
Why did you choose this label convention? It's not really clear to me what `location` means in this context. Also given the `image_scan` prefix of the metric, a label value of `image-scan` seems redundant. What are you trying to communicate? I think there is probably a better name.

jschnath replied:
The `location` would be where the scan semaphore is located, since we have  multiple different ones (in image scan, image enricher, and local scan). Agree with the prefix of the metric, it should just be scan probably, renamed it to that. If you have a better idea than `location` I'm up for it.


Regarding the other comment I've done plenty of testing and some improvement of the metrics and I think the data is correct now.
(Sorry for the late response, I was debugging the metrics and also occupied with something else.)

stehessel replied:
I'm not a native speaker, but afaik `location` suggests a physical location - for example, I'm located in Germany. That doesn't make sense in this context about scanners. Also I think the `image_scan` prefix is fine, but not as a label value. I'd suggest something like `entity={central-service|enricher|sensor}`.

stehessel reviewed (commented):
> I suppose the best way to test this is to put the image on a cluster and trigger image scans to see if the metric reports expected results, but I have not yet done that.

Is there a reason you have not done this? That's exactly what I would have done to test the metrics.

vikin91 commented:
@jschnath is it ready for another look? Maybe you could address the red CI checks first and ping us to review again there?

jschnath commented:
@vikin91 Yeah the metrics look alright now, I'll address the red tests.

eg:

Image enricher:
Held Semaphores:
<img width="1597" alt="image" src="https://github.com/user-attachments/assets/88889334-a4ac-4290-9756-17702bcccae3" />

Semaphore Queue:
<img width="1265" alt="image" src="https://github.com/user-attachments/assets/b63de30f-be27-488c-9346-cacedba3a16f" />


Image scan:
Held semaphores:
<img width="1583" alt="image" src="https://github.com/user-attachments/assets/7bf2d192-bdac-46b7-a041-455cd61ab42e" />

Semaphore Queue:
<img width="1362" alt="image" src="https://github.com/user-attachments/assets/3a23a270-60fa-4043-970b-ac400e6ee175" />

The Image scan data with a some datapoints missing, not sure how that happens. But I think I fixed the bug in the metric logic at least, I'll address the failing tests then mark this for re-review

jschnath commented:
/retest

jschnath commented:
/retest

stehessel reviewed (commented):
See my comment about the metric labels, I think they could be improved. Otherwise looks good to me.

jschnath commented on central/image/service/service_impl.go:
```diff
@@ -86,6 +88,8 @@ var (
 
 	delegateScanPermissions = []string{"Image"}
+
+	imageScanMetricsName = prometheus.Labels{"entity": "central-service"}
```
True, changing it

vikin91 commented on central/image/service/service_impl.go:
```diff
@@ -476,9 +480,17 @@ func (s *serviceImpl) GetImageVulnerabilitiesInternal(ctx context.Context, reque
 func (s *serviceImpl) acquireScanSemaphore(ctx context.Context) error {
 	semaphoreCtx, cancel := context.WithTimeout(ctx, maxSemaphoreWaitTime)
 	defer cancel()
+	images.ScanSemaphoreQueueSize.With(imageScanMetricsLabel).Inc()
+	defer images.ScanSemaphoreQueueSize.With(imageScanMetricsLabel).Dec()
+	defer images.ScanSemaphoreHoldingSize.With(imageScanMetricsLabel).Inc()
```
I am afraid, I do not fully understand the case here. No matter what happens later (due to `defer`), we are already now decreasing the queue size and increase the holding size?
And what happens if `s.internalScanSemaphore.Acquire(semaphoreCtx, 1)` returns an error and the function exits? The number would be incorrect, right?

vikin91 replied:
I changed the logic in [4c91c8c](https://github.com/stackrox/stackrox/pull/14695/commits/4c91c8c3a88bc31d779ed09141fb194f05cba360) - resolving.

vikin91 commented on pkg/images/metrics.go:
```diff
@@ -0,0 +1,33 @@
+		Namespace: metrics.PrometheusNamespace,
+		Subsystem: metrics.CentralSubsystem.String(),
+		Name:      "image_scan_semaphore_queue_size",
+		Help:      "A gauge vector that tracks the size of the queue for the scan semaphores used in scans.",
```
Let me rephrase that sentence the way I understand it and you tell me whether this is correct, ok?

```suggestion
		Help:      "A gauge that shows how many scan tasks are currently waiting for the semaphore to be available.",
```

jschnath replied:
semaphores, plural, but yes, done

vikin91 commented on pkg/images/metrics.go:
```diff
@@ -0,0 +1,33 @@
+		Namespace: metrics.PrometheusNamespace,
+		Subsystem: metrics.CentralSubsystem.String(),
+		Name:      "image_scan_semaphore_limit",
+		Help:      "A gauge vector that tracks the limit of available scan semaphores.",
```
```suggestion
		Help:      "A gauge that shows the maximum number of available scan semaphores. It is configured through env and can only change on pod restart.",
```

jschnath replied:
👍

vikin91 reviewed (dismissed):
Thanks for improving! I spotted one place where I believe we don't do the metrics calculation correctly. Could you check, if my understanding is correct?

jschnath commented on sensor/common/scan/scan.go:
```diff
@@ -146,9 +150,11 @@ func (s *LocalScan) EnrichLocalImageInNamespace(ctx context.Context, centralClie
+	if err := acquireSemaphore(semaphoreCtx, scanLimitSemaphore); err != nil {
 		return nil, errors.Join(err, ErrTooManyParallelScans, ErrEnrichNotStarted)
 	}
+	images.ScanSemaphoreHoldingSize.With(sensorScanMetricsLabel).Inc()
+	defer images.ScanSemaphoreHoldingSize.With(sensorScanMetricsLabel).Dec()
```
Yeah, I think I made it more consistent now, thanks

jschnath commented:
/retest

vikin91 commented on pkg/images/metrics.go:
```diff
@@ -0,0 +1,34 @@
+	}, []string{"entity"})
+	ScanSemaphoreHoldingSize = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: metrics.PrometheusNamespace,
+		Subsystem: metrics.CentralSubsystem.String(),
```
This issue should be fixed in: [d8916e0](https://github.com/stackrox/stackrox/pull/14695/commits/d8916e06384643ab81c039c02997838ce265e457)

vikin91 replied:
@stehessel I have made another pass and fixed the issues that I discovered last time we spoke. Do you maybe want to check ^^ this fix for the issue and the other one in https://github.com/stackrox/stackrox/pull/14695#discussion_r2129178924?

vikin91 commented on central/image/service/service_impl.go:
```diff
@@ -499,6 +511,8 @@ func (s *serviceImpl) GetImageVulnerabilitiesInternal(ctx context.Context, reque
 func (s *serviceImpl) acquireScanSemaphore(ctx context.Context) error {
 	semaphoreCtx, cancel := context.WithTimeout(ctx, maxSemaphoreWaitTime)
 	defer cancel()
+	images.ScanSemaphoreQueueSize.With(imageScanMetricsLabel).Inc()
+	defer images.ScanSemaphoreQueueSize.With(imageScanMetricsLabel).Dec()
```
...which is exactly what this metric tracks. It counts how many requests are waiting in queue to acquire a semaphore. 

The short spike could theoretically be avoided by first doing a pre-check whether the required number of semaphores is available and then acquiring them in a second step - only if there are not enough sema' available, we would then inc the metric. But that has pitfalls - what if the number of available semaphores changes between the two calls? We would need to do a mutex lock and wrap the check and acquire in a transaction. That feels too complex for a small improvement.

parametalol replied:
Right. I forgot about the fact that the acquisition might not be instantaneous.

vikin91 commented on central/image/service/service.go:
```diff
@@ -44,6 +45,7 @@ func New(
 	scanWaiterManager waiter.Manager[*storage.Image],
 	clusterSACHelper sachelper.ClusterSacHelper,
 ) Service {
+	images.ScanSemaphoreLimit.With(imageScanMetricsLabel).Set(float64(env.MaxParallelImageScanInternal.IntegerSetting()))
```
Jan created the initial implementation with the scope of image scanning. That happens in Central and Sensor, so the metric is defined in `pkg` as a shared place. I remember Jan having style-checks failed when he tried cross-importing from central code to sensor code.

Does this answer you question? I am not fully sure what is the relation of the question to the line being commented on, as the `central/metrics/central.go` contains metric definitions and this file sets the metric value.

parametalol replied:
The only arguable benefit is hiding the low-level Prometheus API from the business logic, and making a single place where all the metrics are listed. The list of labels is also created in one place, and with no helpers the business logic is left with no help.
I think at least `images.ScanSemaphoreLimit.With(imageScanMetricsLabel)` could be wrapped somehow.
I'm fine with not implementing any common helpers in this PR.

vikin91 commented on sensor/common/scan/scan.go:
```diff
@@ -47,6 +49,17 @@ var (
+		"entity":    "sensor-active-image-scan",
+	}
+	// sensorActiveScanMetricsLabel is used when collecting metrics about the semaphore for delegated image scans
+	sensorAdHocScanMetricsLabel = prometheus.Labels{
```
[4ab9821](https://github.com/stackrox/stackrox/pull/14695/commits/4ab982189eadd5bc67bffdaa2236a888c1c2cdb9)

parametalol commented on central/image/metrics/metrics.go:
```diff
@@ -0,0 +1,19 @@
+		Namespace: metrics.PrometheusNamespace,
+		Subsystem: metrics.CentralSubsystem.String(),
+		Name:      "image_scan_semaphore_queue_size",
+		Help:      "A counter that tracks the queue size of the scan semaphre used in image scan.",
```
I reviewed a version with a typo, apparently. Now it's gone. Thanks.

vikin91 commented on sensor/common/scan/scan.go:
```diff
@@ -47,6 +49,17 @@ var (
 	log = logging.LoggerForModule()
+
+	// sensorActiveScanMetricsLabel is used when collecting metrics about the semaphore for active image scans
+	sensorActiveScanMetricsLabel = prometheus.Labels{
```
That may be the case. Let me then propose a verbose name without introducing new words like "active".

dcaravel commented on sensor/common/scan/scan.go:
```diff
@@ -137,19 +157,27 @@ func (s *LocalScan) EnrichLocalImageInNamespace(ctx context.Context, centralClie
 
 	semaphoreCtx, cancel := context.WithTimeout(ctx, s.maxSemaphoreWaitTime)
 	defer cancel()
+	images.ScanSemaphoreQueueSize.With(metricLabels).Inc()
```
Similar comment `ScanSemaphoreQueueSize` will not be decremented if `Acquire` fails, or am i missing something?

dcaravel commented on sensor/common/scan/scan.go:
```diff
@@ -47,6 +49,19 @@ var (
 
 	log = logging.LoggerForModule()
+
+	// sensorCentralScanMetricsLabel is used when collecting metrics about the semaphore for image scans
+	// that are triggered by sending a request to scanner over central.
+	sensorCentralScanMetricsLabel = prometheus.Labels{
+		"subsystem": "sensor",
+		"entity":    "image-scan-over-central",
+	}
+	// sensorAdHocScanMetricsLabel is used when collecting metrics about the semaphore for delegated image scans
+	// that are triggered by sending a request to the scanner instance running in the secured cluster.
+	sensorAdHocScanMetricsLabel = prometheus.Labels{
+		"subsystem": "sensor",
+		"entity":    "delegated-image-scan",
+	}
```
Let's clean up the verbiage here - everything in `scan.go` is for handling a 'delegated scan'. If Sensor is handling the scan, it is a 'delegated scan'.

The distinction here is for what is requesting the scan, was it Central (from adhoc requests via roxctl, api, etc.) or Sensor (from observed k8s workloads, reprocessing, etc.). 

scanSemaphore = Delegated scans requested by Sensor
adHocScanSemaphore = Delegated scans requested by Central

vikin91 replied:
Thanks for that clarification. I added one more label to track the origin of the request in [8b93928](https://github.com/stackrox/stackrox/pull/14695/commits/8b93928c3a492ae09ef77a64488daa9438d8376c)

vikin91 commented on pkg/images/enricher/enricher_impl.go:
```diff
@@ -965,11 +971,18 @@ func (e *enricherImpl) enrichImageWithScanner(ctx context.Context, image *storag
 	}
 
 	sema := scanner.MaxConcurrentScanSemaphore()
+	images.ScanSemaphoreQueueSize.With(enricherScanMetricsLabel).Inc()
```
Good catch! Pls check [0401809](https://github.com/stackrox/stackrox/pull/14695/commits/04018095f78715c0dbb3f9b2078b20c469800081) and [e0a6ba3](https://github.com/stackrox/stackrox/pull/14695/commits/e0a6ba3687a704e626fa505852789f41313c6eb6)

vikin91 commented:
@stehessel @dcaravel, I have addressed all open comments. I have resolved some comments of Stephan due to inactivity cause I didn't know he was on PTO. Let me know if you want to have another look. If not, I will merge this soon.
